pulsar-transformations/src/main/java/com/datastax/oss/pulsar/functions/transforms/jstl/JstlFunctions.java

```diff
@@ -67,33 +66,13 @@ public static long now() {
     return clock.millis();
   }
 
-  public static long dateadd(Object input, long delta, String unit) {
-    if (input instanceof String) {
-      return dateadd((String) input, delta, unit);
-    } else if (input instanceof Long) {
-      return dateadd((long) input, delta, unit);
+  public static long dateadd(Object input, Object delta, Object unit) {
+    if (input == null || unit == null) {
+      throw new ELException(LocalMessages.get("error.method.notypes"));
     }
 
-    throw new IllegalArgumentException(
-        "Invalid input type: "
-            + input.getClass().getSimpleName()
-            + ". Should either be an RFC3339 datetime string like '2007-12-01T12:30:00Z' or epoch millis");
-  }
-
-  private static long dateadd(String input, long delta, String unit) {
-    OffsetDateTime offsetDateTime = OffsetDateTime.parse(input);
-    return dateadd(offsetDateTime, delta, unit);
-  }
-
-  private static long dateadd(long epochMillis, long delta, String unit) {
-    Instant instant = Instant.ofEpochMilli(epochMillis);
-    OffsetDateTime localDateTime = OffsetDateTime.ofInstant(instant, ZoneOffset.UTC);
-    return dateadd(localDateTime, delta, unit);
-  }
-
-  private static long dateadd(OffsetDateTime offsetDateTime, long delta, String unit) {
-    final ChronoUnit chronoUnit;
-    switch (unit) {
+    ChronoUnit chronoUnit;
+    switch (CustomTypeConverter.INSTANCE.coerceToString(unit)) {
```

Comment:
this way, `unit` is more permissive than expected for this use case. what's wrong with limiting it `String` type in the method signature and just using it as is?

Reply:
It's not really more permissive, the values allowed are still the same. It just allows to use a type that is compatible.
Practically, only BYTES can match.